### ISSUE TITLE
Zest does not calculate page length correctly in assertion

### ISSUE THREAD

**Issue by g-sato** [zaproxy/zaproxy#2669]
Hello zap team,

I've found a bug about that ZAP do not calculate page length correctly in case of website contains multibyte text when assert length in ZEST Script.

［Steps to reproduce］

1) Download this file, and rename file multibyte.html.txt to multibyte.html.
(This html file containts Japanese Text 'こんにちは' means 'Hello' in English)

[multibyte.html.txt](https://github.com/zaproxy/zaproxy/files/359719/multibyte.html.txt)

2) Put multibyte.html to web server.

3) Run ZAP 2.5.0 and Access multibyte.html on web server.

4) Create a new Zest standalone script, and add the history of multibyte.html to that Zest Script.

After this, two assertion added automatically below:
'GET multibyte.html' step that 'Assert - Status Code(200)'
'Assert - Length(response body = 58 +/- 1%)'.

5) Run that Zest standalone script under two condition below:

CASE1:
Response header contains 
>  Content-Type: text/html; charset=UTF-8

CASE2:
Response header contains 
>  Content-Type: text/html

RESULT:

CASE1:

```
HTTP/1.1 200 OK
Date: Tue, 12 Jul 2016 15:06:36 GMT
Server: Apache/2.4.12 (Win32) OpenSSL/1.0.1l PHP/5.6.8
Last-Modified: Tue, 12 Jul 2016 09:16:27 GMT
ETag: "3a-5376cb937f0c0"
Accept-Ranges: bytes
Content-Length: 58
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8

<html>
<body>
こんにちは(Hello)
</body>
</html>
```

-> Assertion of Content Length returns 'FAILED'. Because Content-Length is 48.

> 6	GET	http://localhost/zaptest/multibyte.html	200	OK	14307	58	jar:file:/C:/Users/user123/Desktop/ZAP_2.5.0/zap-2.5.0.jar!/resource/icon/16/101.png	FAILED Assert - response.body length: expected 58 got 48 (17% difference)

CASE2:

```
HTTP/1.1 200 OK
Date: Tue, 12 Jul 2016 15:07:34 GMT
Server: Apache/2.4.12 (Win32) OpenSSL/1.0.1l PHP/5.6.8
Last-Modified: Tue, 12 Jul 2016 09:16:27 GMT
ETag: "3a-5376cb937f0c0"
Accept-Ranges: bytes
Content-Length: 58
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html

<html>
<body>
こんにちは(Hello)
</body>
</html>

```
(the same response except Content-Type header.)
-> Assertion of Content Length returns 'OK'. Content-Length is 58.

> 9	GET	http://localhost/zaptest/multibyte.html	200	OK	2022	58	jar:file:/C:/Users/user123/Desktop/ZAP_2.5.0/zap-2.5.0.jar!/resource/icon/16/102.png	

[Additional Information]
When test webpage containts singlebyte text, this bug is not represent.

Regards.

**Comment by kingthorin** [comment]
What happens when the content type is `Content-Type: text/html; charset=UTF-16`?

**Comment by g-sato** [comment]
Hi, Answers follow:

Q1:
> What happens when the content type is Content-Type: text/html; charset=UTF-16?

```
HTTP/1.1 200 OK
Date: Tue, 12 Jul 2016 20:15:53 GMT
Server: Apache/2.4.12 (Win32) OpenSSL/1.0.1l PHP/5.6.8
Last-Modified: Tue, 12 Jul 2016 09:16:27 GMT
ETag: "3a-5376cb937f0c0"
Accept-Ranges: bytes
Content-Length: 58
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-16

xxxxxxxxx(* Response body is garbled because multibyte text encoding is UTF-8)
```

-> Assertion of Content Length returns 'FAILED'. Because Content-Length is 29.

> 17	GET	http://localhost/zaptest/multibyte.html	200	OK	8318	60	jar:file:/C:/Users/user123/Desktop/ZAP_2.5.0/zap-2.5.0.jar!/resource/icon/16/101.png	FAILED Assert - response.body length: expected 58 got 29 (50% difference)

Q2:

> Also what version of the Zest addon are you using?

My Zest addon version is 21 (newest).

Regards.

**Comment by thc202** [comment]
This issue affects older versions of ZAP and Zest add-on.

The initial value of the assertion is the size of the response body in bytes [1], while the value used to verify the assertion is the number of chars [2], which might not be the same depending on the character encoding being used.

I guess changing the former to also use chars is the/an appropriate fix?

[1] https://github.com/zaproxy/zap-extensions/blob/ad229c6d69c1ceb9ff1a36bb7dfcb87f74e172ab/src/org/zaproxy/zap/extension/zest/ExtensionZest.java#L696
[2] https://github.com/mozilla/zest/blob/635aa9a38188c5578ca7a5d1859959d4516051dc/src/org/mozilla/zest/core/v1/ZestExpressionLength.java#L129

**Comment by thc202** [comment]
Fixed in zaproxy/zap-extensions#458.

**Comment by thc202** [comment]
Released in version 22 of Zest add-on.
